File: ProgrammingLanguage/1/20170517_1.py
import time
import math

numCnt = None
nums = [0]
isPrime = [0, 0]
startClock = None
endClock = None


def GCD(a, b):
    if (a < b):
        a, b = b, a

    while (True):
        r = a % b
        if (r == 0):
            return b
        a, b = b, r


def eratos():
    global isPrime
    isPrime = [0, 0]
    isPrime = isPrime + [True for i in range(100000-1)]
    for i in range(2, math.ceil((100000+1)**0.5)):
        if (isPrime[i]):
            for j in range(i*i, 100000+1, i):
                isPrime[j] = False


def main():

    print("Input the number of numbers to process: ")
    numCnt = 0
    while(True):
        try:
            numCnt = int(input())
            if(not (2 <= numCnt and numCnt <= 30)):
                raise Exception()
        except Exception as e:
            print("Invalid input. Try again: ")
            continue
        break

    print("Input the numbers to be processed:")
    while(True):
        try:
            nums = list(map(int, input().split()))
            for item in nums:
                if(not(1 <= item and item <= 1000000)):
                    raise Exception()
            if(len(nums) != numCnt):
                raise Exception()
            nums = [0] + nums
        except Exception as a:
            print("Invalid input. Try again: ")
            continue
        break

    startClock = time.time()

    gcd = GCD(nums[1], nums[2])
    for i in range(3, numCnt+1):
        gcd = GCD(nums[i], gcd)
    print("GCD of input numbers is " + str(gcd))

    for i in range(1, numCnt):
        minIdx = i
        tmp = nums[i]
        for j in range(i+1, numCnt+1):
            if(nums[minIdx] > nums[j]):
                minIdx = j

        nums[i] = nums[minIdx]
        nums[minIdx] = tmp

    eratos()
    global isPrime
    for i in range(1, numCnt):
        cnt = 0
        if(nums[i] == nums[i+1]):
            continue

        for j in range(nums[i], nums[i+1]+1):
            if(isPrime[j]):
                cnt += 1
        print("Number of prime numbers between" +
              str(nums[i]) + ", "+str(nums[i+1])+": " + str(cnt))
    endClock = time.time()
    duration = (endClock-startClock)
    print("Total execution time using Python is "+str(duration)+" seconds!")


main()

File: ProgrammingLanguage/1/20170517_1.cpp
#include <ctime>
#include <iostream>
#include <limits.h>

using namespace std;

int numCnt;
int nums[31];
bool isPrime[100001] = {0, 0};
clock_t startClock;
clock_t endClock;

int GCD(int a, int b) {
  if (a < b) {
    int tmp = a;
    a = b;
    b = tmp;
  }
  while (true) {
    int r = a % b;
    if (r == 0)
      return b;
    a = b;
    b = r;
  }
}

void eratos() {
  for (int i = 2; i <= 100000; i++) {
    isPrime[i] = true;
  }
  for (int i = 2; i * i <= 100000; i++) {
    if (isPrime[i]) {
      for (int j = i * i; j <= 100000; j += i) {
        isPrime[j] = false;
      }
    }
  }
}

int main() {
  cout << "Input the number of numbers to process: ";
  numCnt = 0;

  while (!(cin >> numCnt) || !(2 <= numCnt && numCnt <= 30)) {
    cin.clear();
    cin.ignore(INT_MAX, '\n');
    cout << "Invalid input.  Try again: ";
  }

  cout << "Input the numbers to be processed:" << endl;
  bool isInputOk = true;
  int i = 1;
  while (i <= numCnt) {
    isInputOk = true;
    while (!(cin >> nums[i]) || !(1 <= nums[i] && nums[i] <= 100000)) {
      cin.clear();
      cin.ignore(INT_MAX, '\n');
      cout << "Invalid input.  Try again: ";
      isInputOk = false;
      break;
    }
    i++;
    if (!isInputOk) {
      i = 1;
    }
  }

  startClock = clock();

  int gcd = GCD(nums[1], nums[2]);
  for (int i = 3; i <= numCnt; i++) {
    gcd = GCD(nums[i], gcd);
  }
  cout << "GCD of input numbers is " << gcd << endl;

  for (int i = 1; i < numCnt; i++) { // selection sort
    int minIdx = i;
    int tmp = nums[i];
    for (int j = i + 1; j <= numCnt; j++) {
      if (nums[minIdx] > nums[j]) {
        minIdx = j;
      }
    }
    nums[i] = nums[minIdx];
    nums[minIdx] = tmp;
  }

  eratos();
  for (int i = 1; i < numCnt; i++) {
    int cnt = 0;
    if (nums[i] == nums[i + 1]) {
      continue;
    }
    for (int j = nums[i]; j <= nums[i + 1]; j++) {
      if (isPrime[j]) {
        cnt++;
      }
    }
    cout << "Number of prime numbers between " << nums[i] << ", " << nums[i + 1]
         << ": " << cnt << endl;
  }
  endClock = clock();
  double duration = (double)(endClock - startClock) / CLOCKS_PER_SEC;
  cout << "Total execution time using C++ is " << duration << " seconds!"
       << endl;
}
